### PR TITLE
SNOW-582820 Finer-grained usage telemetry for Snowpark Python

### PR DESCRIPTION
**Pull request by sfc-gh-kwagner**
Description
Added more usage telemetry to snowpark-python

Testing

### CODE REVIEW

File: src/snowflake/snowpark/dataframe.py [hunk 1]
```diff
@@ -1124,12 +1156,22 @@ def drop_duplicates(self, *subset: Union[str, Iterable[str]]) -> "DataFrame":
...
         )
         rownum_name = generate_random_alphanumeric()
-        return (
+        df = (
             self.select(*output_cols, rownum.as_(rownum_name))
             .where(col(rownum_name) == 1)
             .select(output_cols)
         )
+        # Reformat the extra API calls
+        df._plan.api_calls = [
```
[the remark this answers is not in the document]
Comment (sfc-gh-kwagner) [reply src/snowflake/snowpark/dataframe.py:1165]: Sure, they're not dropped, but reclassified as subcalls of this method

**Comment by sfc-gh-kwagner** [comment]
> I'm a bit worried that if we need to to do something like:
> 
> ```
>         res_df._plan.api_calls = [
>             *self._df._plan.api_calls,
>             {
>                 TelemetryField.NAME.value: "DataFrameStatFunctions.sample_by",
>                 TelemetryField.KEY_SUBCALLS.value: res_df._plan.api_calls.copy(),
>             },
>         ]
> ```
> 
> Then it feels a bit hard to maintain when new methods are added. I don't have a good solution to it yet though :(

Yea implementing this for the stats functions was not ideal. Some of these functions don't even return DataFrames, so its not as easy to implement this in a decorator as it is for the DataFrame class.

I'm open to suggestions on this, but don't see a way around this unless we refactor these functions which I thought would be a much more invasive change, but might be the best option.

I can work on that when I'm back from break

**Comment by sfc-gh-jdu** [comment]
> > I'm a bit worried that if we need to to do something like:
> > ```
> >         res_df._plan.api_calls = [
> >             *self._df._plan.api_calls,
> >             {
> >                 TelemetryField.NAME.value: "DataFrameStatFunctions.sample_by",
> >                 TelemetryField.KEY_SUBCALLS.value: res_df._plan.api_calls.copy(),
> >             },
> >         ]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Then it feels a bit hard to maintain when new methods are added. I don't have a good solution to it yet though :(
> 
> Yea implementing this for the stats functions was not ideal. Some of these functions don't even return DataFrames, so its not as easy to implement this in a decorator as it is for the DataFrame class.
> 
> I'm open to suggestions on this, but don't see a way around this unless we refactor these functions which I thought would be a much more invasive change, but might be the best option.
> 
> I can work on that when I'm back from break

But for those methods returning a dataframe, we can get rid of it, right? I still suggest to avoid doing these for methods if possible.

**Comment by sfc-gh-sfan** [comment]
I'm wondering if a quick brainstorming would be helpful. I'm no python expert but maybe other people could contribute and see if there is a better way.

**Comment by sfc-gh-kwagner** [comment]
> I'm wondering if a quick brainstorming would be helpful. I'm no python expert but maybe other people could contribute and see if there is a better way.

Yes lets do that. I'll setup a meeting for tomorrow morning

**Comment by sfc-gh-kwagner** [comment]
I ran some performance tests at @sfc-gh-yixie 's request to measure the speed loss and increase in memory usage of this PR. Here is the code that I ran to do the performance tests:

```
ZERO_DEPTH_BASES = (str, bytes, Number, range, bytearray)


def getsize(obj_0):
    """Recursively iterate to sum size of object & members."""
    _seen_ids = set()
    def inner(obj):
        try:
            obj_id = id(obj)
            if obj_id in _seen_ids:
                return 0
            _seen_ids.add(obj_id)
            size = sys.getsizeof(obj)
            if isinstance(obj, ZERO_DEPTH_BASES):
                pass # bypass remaining control flow and return
            elif isinstance(obj, (tuple, list, Set, deque)):
                size += sum(inner(i) for i in obj)
            elif isinstance(obj, Mapping) or hasattr(obj, 'items'):
                size += sum(inner(k) + inner(v) for k, v in getattr(obj, 'items')())
            # Check for custom object instances - may subclass above too
            if hasattr(obj, '__dict__'):
                size += inner(vars(obj))
            if hasattr(obj, '__slots__'): # can have __slots__ with __dict__
                size += sum(inner(getattr(obj, s)) for s in obj.__slots__ if hasattr(obj, s))
            return size
        except:
            return 0
    return inner(obj_0)


def test_perf_test(session):
    df = session.create_dataframe([[1, 2], [1, 3], [4, 4]], schema=["a", "b"])
    time_to_creates = list()
    time_end_to_ends = list()
    sizes = list()
    sizes_of_api_calls = list()
    num_runs = 10
    num_api_calls = 50
    for i in range(num_runs):
        print(f"\nStarting perf test run {i+1}\n")
        start = time.perf_counter()
        df_new = df
        # Call the withColumns API a certain number of times
        for i in range(num_api_calls):
            df_new = df_new.withColumn(f"c{i}", df["a"])
        time_after_create = time.perf_counter()
        res = df_new.collect()
        end = time.perf_counter()
        size = getsize(df_new)
        #size_of_api_calls = getsize(df_new._plan.api_calls)
        time_to_create = time_after_create - start
        time_end_to_end = end - start
        print(f"Time taken to create end DataFrame: {time_to_create}")
        print(f"Time taken to execute end-to-end: {time_end_to_end}")
        print(f"Size of DataFrame: {size}")
        sizes.append(size)
        #sizes_of_api_calls.append(size_of_api_calls)
        time_to_creates.append(time_to_create)
        time_end_to_ends.append(time_end_to_end)

    print(f"\n\nAvg. Time Taken to create end DataFrame: {sum(time_to_creates) / num_runs}")
    print(f"Avg. Time Taken to execute end-to-end DataFrame: {sum(time_end_to_ends) / num_runs}")
    print(f"Avg. DataFrame Size: {sum(sizes) / num_runs}")
    print(f"Avg. api_calls Size: {sum(sizes_of_api_calls) / num_runs}")
```
Just uncomment this #sizes_of_api_calls.append(size_of_api_calls) and this #size_of_api_calls = getsize(df_new._plan.api_calls) when running with the telemetry branch.

Here are the results that I got. As you can see, the telemetry PR is slightly faster than main and the size of the data that it uses becomes smaller (as a percentage) as more APIs are called (since it takes up 2.89% of the objects total space when running with 50 APIs and then 1.61% when running with 100 APIs). Also, the size of api_calls increases linearly, but the size of the DataFrame object increases by n^2

```
Branch - Test Runs - Number of withColumn API calls

Main - 5 - 50
Avg. Time Taken to create end DataFrame: 7.915868553400287
Avg. Time Taken to execute end-to-end DataFrame: 8.281638565600042
Avg. DataFrame Size: 1699248.6

Main - 10 - 50
Avg. Time Taken to create end DataFrame: 7.88218994259978
Avg. Time Taken to execute end-to-end DataFrame: 8.219733467999868
Avg. DataFrame Size: 1700509.5

Main - 10 - 100
Avg. Time Taken to create end DataFrame: 13.801145379900026
Avg. Time Taken to execute end-to-end DataFrame: 14.75949425050003
Avg. DataFrame Size: 6306070.9

Telemetry - 5 - 50
Avg. Time Taken to create end DataFrame: 7.753432376400087
Avg. Time Taken to execute end-to-end DataFrame: 8.118983916200069
Avg. DataFrame Size: 1814859.2
Avg. api_calls Size: 52483.0 (takes up 2.89% of total space)

Telemetry - 10 - 50
Avg. Time Taken to create end DataFrame: 7.535788094700001
Avg. Time Taken to execute end-to-end DataFrame: 7.886025032100042
Avg. DataFrame Size: 1816852.3
Avg. api_calls Size: 52483.0 (takes up 2.89% of total space)

Telemetry - 10 - 100
Avg. Time Taken to create end DataFrame: 13.719506314400041
Avg. Time Taken to execute end-to-end DataFrame: 14.752738628999941
Avg. DataFrame Size: 6505397.0
Avg. api_calls Size: 104431.0 (takes up 1.61% of total space)
```

**Review by sfc-gh-sfan — APPROVED** [review]

**Pull request merged**